### PR TITLE
adding optional D4 support

### PR DESCRIPTION
**Pull request by jpoto**
This pull request contains the optional implementation of D4 from Grimme

For now it is deactivated per default. 

@oschuett 
Can you check the build system? 
I set it up that it downloads ninja and the D4 package from the public repository, not a copy in https://www.cp2k.org/static/downloads

@fstein93 
Maybe you can have a look?

### CODE REVIEW

File: src/qs_dispersion_pairpot.F [hunk 1]
```diff
@@ -2006,6 +2041,15 @@ SUBROUTINE calculate_dispersion_pairpot(qs_env, dispersion_env, energy, calculat
                   END DO
 
                END IF
+            ELSE IF (disp_a%type == dftd4_pp .AND. dr > 0.001_dp) THEN
+#if defined(__DFTD4)
```
Comment (fstein93) [inline src/qs_dispersion_pairpot.F]: Same here

File: src/qs_dispersion_pairpot.F [hunk 2]
```diff
@@ -1368,6 +1387,26 @@ SUBROUTINE calculate_dispersion_pairpot(qs_env, dispersion_env, energy, calculat
          CALL get_atomic_kind_set(atomic_kind_set, atom_of_kind=atom_of_kind, kind_of=kind_of)
       END IF
 
+      evdw = 0._dp
+      pv_virial_thread(:, :) = 0._dp
+      IF (dispersion_env%pp_type == vdw_pairpot_dftd4) THEN
+#if defined(__DFTD4)
```
Comment (fstein93) [inline src/qs_dispersion_pairpot.F]: Please move this code to a separate module.

File: src/qs_dispersion_pairpot.F [hunk 3]
```diff
@@ -3626,4 +3670,106 @@ SUBROUTINE exclude_d3_kind_pair(exclude_list, za, zb, zc, exclude)
 
    END SUBROUTINE exclude_d3_kind_pair
 
+#if defined(__DFTD4)
```
Comment (fstein93) [inline src/qs_dispersion_pairpot.F]: and here

**Review by fstein93 — COMMENTED** [review]
Please move the DFTD4-related code to a separate module.

**Comment by fstein93** [comment]
Reduce the size of the test systems to something which runs within a few seconds.

File: src/qs_dispersion_d4.F [hunk 4]
```diff
@@ -0,0 +1,170 @@
+!--------------------------------------------------------------------------------------------------!
+!   CP2K: A general program to perform molecular dynamics simulations                              !
+!   Copyright 2000-2024 CP2K developers group <https://cp2k.org>                                   !
+!                                                                                                  !
+!   SPDX-License-Identifier: GPL-2.0-or-later                                                      !
+!--------------------------------------------------------------------------------------------------!
+
+! **************************************************************************************************
+!> \brief Calculation of dispersion using pair potentials
+!> \author JGH
```
Comment (fstein93) [inline src/qs_dispersion_d4.F]: Please use your own name.

File: src/qs_dispersion_d4.F [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+! **************************************************************************************************
+MODULE qs_dispersion_d4
+
+   USE atomic_kind_types,               ONLY: get_atomic_kind
+   USE cell_types,                      ONLY: cell_type,&
+                                              get_cell
+   USE dftd4,                           ONLY: d4_model,&
```
Comment (fstein93) [inline src/qs_dispersion_d4.F]: You have to wrap this with `#if defined(__DFTD4) ... #endif`

File: src/qs_dispersion_d4.F [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+!> \param virial ...
+! **************************************************************************************************
+   SUBROUTINE calculate_dispersion_d4_pairpot(dispersion_env, particle_set, cell, energy, force, atom_of_kind, virial)
+
+      class(damping_param), ALLOCATABLE                  :: param
+      TYPE(d4_model)                                     :: disp
+      TYPE(structure_type)                               :: mol
+      TYPE(realspace_cutoff)                             :: cutoff
+
+      TYPE(qs_dispersion_type), POINTER, INTENT(IN)                       :: dispersion_env
```
Comment (fstein93) [inline src/qs_dispersion_d4.F]: Move the block with the declarations of dummy variables upwards. That's how it is the case in all other parts of the code. Don't forget to run `make pretty` again.

File: src/qs_dispersion_d4.F [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+         POINTER                                         :: particle_set
+      TYPE(cell_type), INTENT(IN), POINTER               :: cell
+      REAL(KIND=dp), INTENT(OUT)                         :: energy
+      TYPE(qs_force_type), DIMENSION(:), INTENT(INOUT), &
+         OPTIONAL, POINTER                               :: force
+      INTEGER, DIMENSION(:), INTENT(IN), OPTIONAL        :: atom_of_kind
+      REAL(KIND=dp), DIMENSION(3, 3), INTENT(OUT), &
+         OPTIONAL                                        :: virial
+
+      CPABORT("Build without DFTD4")
```
Comment (fstein93) [inline src/qs_dispersion_d4.F:188]: Add `MARK_USED(variable)` for all dummy variables here to make the compiler(s) happy if DFTD4 is not in use.

**Comment by fstein93** [comment]
@oschuett Do you agree on adding ninja to the dependencies?

**Comment by oschuett** [comment]
I'm fine with adding ninja to the toolchain.
I've put copies on our server so you can use `download_pkg_from_cp2k_org`:
- https://www.cp2k.org/static/downloads/ninja-v1.12.1.tar.gz
- https://www.cp2k.org/static/downloads/dftd4-3.6.0.tar.gz

**Comment by fstein93** [comment]
@oschuett Shall it still be included in this release 2024.2 or with the release 2025.1?

**Comment by oschuett** [comment]
Since this PR seems almost ready to be merged, I see no problem to include it in the upcoming 2024.2 release.

**Pull request merged**